### PR TITLE
Adding reverse geocoding address launcher widget

### LINKED ISSUE #628: Geocoding-related app widget example
Some how combine location/maps with an app widget?

https://developer.android.com/guide/topics/appwidgets/index.html

https://medium.com/android-bits/android-widgets-ad3d166458d3

### PR DESCRIPTION
**Pull request by langsmith**
Resolves #628 by adding a launcher screen app widget to the demo app. Using reverse geocoding and the location layer plugin, the device's location/address eventually displays in the widget

screen shot 2018-04-23 at 3 09 14 pm

### CODE REVIEW

**Comment by langsmith** [comment]
ezgif com-resize 1

**Comment by osana** [comment]
Another question or use case:
  Last Location is null, widget is installed, location is requested
I believe provider might quit before the update location gets called.
Do we need to consider this case or we just do not care about this?

Provider should probably wait for the callback from the LocationEngine .

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/DemoAppHomeScreenAddressWidget.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+      .query(Point.fromLngLat(currentDeviceLocation.getLongitude(), currentDeviceLocation.getLatitude()))
+      .build();
+
+    reverseGeocode.enqueueCall(new Callback<GeocodingResponse>() {
+      @Override
+      public void onResponse(Call<GeocodingResponse> call, Response<GeocodingResponse> response) {
+        Log.d(tag, "onResponse: ");
+
+        // Check that the reverse geocoding response has a place name to display
+        if (response.body().features().size() > 0 && !response.body().features().get(0).placeName().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/DemoAppHomeScreenAddressWidget.java]: Fixed

**Comment by langsmith** [comment]
>I believe provider might quit before the update location gets called.

I'm not sure what you mean. Possible to clarify?

**Comment by langsmith** [comment]
Using a LinearLayout for the widget layout, because only certain  layout classes are allowed https://developer.android.com/guide/topics/appwidgets/#CreatingLayout

**Comment by langsmith** [comment]
Ok, I think this ready for another (final) review.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/DemoAppHomeScreenAddressWidget.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+      return null;
+    } else if (context instanceof ContextWrapper) {
+      if (context instanceof Activity) {
+        return (Activity) context;
+      } else {
+        return getActivity(((ContextWrapper) context).getBaseContext());
+      }
+    }
+    return null;
+  }
```
Comment (tobrun) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/DemoAppHomeScreenAddressWidget.java]: getActivity can be removed, only one if case should execute seeing the usages,  can be directly cast if needed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/DemoAppHomeScreenAddressWidget.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+/**
+ * Use the Mapbox Java SDK to make a reverse geocode search with the device's current coordinates. Display
+ * the address in an Android home screen widget.
+ */
+public class DemoAppHomeScreenAddressWidget extends AppWidgetProvider implements
+  PermissionsListener {
+
+  private Context context;
+  private LocationChangeListeningActivityLocationCallback callback;
+  private static final long DEFAULT_INTERVAL_IN_MILLISECONDS = 1000L;
```
[the remark this answers is not in the document]
Comment (langsmith) [reply MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/utils/DemoAppHomeScreenAddressWidget.java]: Fixed

**Pull request merged**